File: backend/.NET/Calindra.Desafio/Calindra.Desafio.Domain/Services/DistanceCalculatorService.cs
using Calindra.Desafio.Domain.Commands.Inputs;
using Calindra.Desafio.Domain.Exceptions;
using Calindra.Desafio.Domain.Models.Responses;
using Calindra.Desafio.Domain.Results;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Threading.Tasks;

namespace Calindra.Desafio.Domain.Services
{
    public class DistanceCalculatorService
    {
        private readonly ILogger<DistanceCalculatorService> _logger;
        private readonly IGeolocationApiService _geolocationApiService;

        public DistanceCalculatorService(ILogger<DistanceCalculatorService> logger, IGeolocationApiService geolocationApiService)
        {
            _logger = logger;
            _geolocationApiService = geolocationApiService;
        }

        public async Task<MethodResult> CalculateDistancesFromAddressess(Addresses addresses)
        {
            try
            {
                var addressesInfo = await RequestAddressesInfo(addresses);
                if (addressesInfo.Count < 2)
                    return Result.FailWithData(nameof(addresses), HttpStatusCode.BadRequest, "There must be at least 2 valid addresses");

                var (shorterQuantity, greatherQuantity) = CalculateQuantities(addresses.AddressList.Count);

                foreach (var point in addressesInfo)
                {
                    foreach (var p in addressesInfo.Where(p => p != point))
                        point.Addresses.Add(new GeolocationInfo(p, CalculateEuclideanDistance(point, p)));

                    point.ShorterDistances = point.Addresses.OrderBy(p => p.Distance).Take(shorterQuantity).ToList();
                    point.GreatherDistances = point.Addresses.OrderByDescending(p => p.Distance).Take(greatherQuantity).ToList();
                }

                return Result.Ok(addressesInfo);
            }
            catch (CustomException e)
            {
                _logger.LogError(e, $"Erro ao processar as distâncias! \nMessage: {e.Message}");
                return Result.FailWithData(string.Empty, HttpStatusCode.BadRequest, "Unable to process distances", $"Message: {e.Message}");
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Erro ao processar as distâncias");
                return Result.Fail(string.Empty, "Unable to process distances");
            }
        }

        private (int shorterQuantity, int greatherQuantity) CalculateQuantities(int quantity)
        {
            quantity -= 1;
            var shorter = (int)(quantity / 2);
            var greather = (int)(quantity / 2) + (quantity % 2);
            return (shorter, greather);
        }

        private async Task<List<GeolocationReferencesInfo>> RequestAddressesInfo(Addresses command)
        {
            var tasks = command.AddressList.Select(async address => await _geolocationApiService.GetGeolocationFrom(address));
            var results = await Task.WhenAll(tasks);

            var points = results
                .Where(result => result.Status == "OK")
                .SelectMany(result => result.Results)
                .Select(result => new GeolocationReferencesInfo
                {
                    PlaceId = result.PlaceId,
                    Address = result.FormattedAddress,
                    Location = new Location(result.Geometry.Location.Lat, result.Geometry.Location.Lng),
                })
                .ToList();

            return points;
        }

        private static double CalculateEuclideanDistance(GeolocationReferencesInfo pointA, GeolocationReferencesInfo pointB)
        {
            return Math.Sqrt(
                Math.Pow(pointA.Location.Lat - pointB.Location.Lat, 2) +
                Math.Pow(pointA.Location.Lng - pointB.Location.Lng, 2)
            );
        }
    }
}

File: backend/Node.JS/calindra-desafio/src/controllers/v1.0/geolocation/services/distance-calculator/distance-calculator.service.ts
import { Injectable } from '@nestjs/common';
import { forkJoin, map, Observable } from 'rxjs';

import { GoogleGeolocationApiService } from '../google-geolocation-api/google-geolocation-api.service';
import { Addresses } from '../../commands/inputs/addresses';
import { GeolocationReferencesInfo } from '../../models/responses/geolocation-references-info';
import { GeolocationResponse } from '../../models/geolocation/geolocation-response';
import { GeolocationResult } from '../../models/geolocation/geolocation-result';
import { Location } from '../../models/responses/location';
import { GeolocationInfo } from '../../models/responses/geolocation-info';
import { MethodResult } from 'src/shared/results/method-result';
import { Result } from 'src/shared/results/result';

@Injectable()
export class DistanceCalculatorService {

    constructor(private geolocationApiService: GoogleGeolocationApiService) { }

    public calculateDistancesFromAddressess(addresses: Addresses): Observable<MethodResult> {
        const { shorterQuantity, greatherQuantity } = this.calculateQuantities(addresses.addressList);
        const referencesInfo = this.requestGeolocationReferencesInfo(addresses);
        return referencesInfo.pipe(
            map((references: GeolocationReferencesInfo[]) => {
                try {
                    references.forEach(reference => {
                        references.filter(ref => ref != reference).forEach(ref => {
                            const geolocationInfo = this.mapGeolocationReferencesInfoToGeolocationInfo(ref);
                            geolocationInfo.distance = this.calculateEuclidianDistance(reference, ref);
                            reference.addresses.push(geolocationInfo);
                        });
                        this.orderByDistanceAndTakeShorthers(reference, shorterQuantity);
                        this.orderByDescendingDistanceAndTakeGreathers(reference, greatherQuantity);
                    });
                    return Result.ok(references);
                } catch (error) {
                    return Result.fail('', 'Ocorreu uma falha no processamento da requisição');
                }
            })
        );
    }

    private requestGeolocationReferencesInfo(addresses: Addresses): Observable<GeolocationReferencesInfo[]> {
        const tasks$ = addresses.addressList.map(address => this.geolocationApiService.getGeolocationFrom(address));
        const results = forkJoin(tasks$);
        return results.pipe(map((gelocationResponses: GeolocationResponse[]) => this.mapGeolocationResponsesToGeolocationReferencesInfo(gelocationResponses)));
    }

    private calculateQuantities(addresses: string[]) {
        const quantity = addresses.length - 1;
        const part = Math.trunc(quantity / 2);
        return {
            shorterQuantity: part,
            greatherQuantity: Math.trunc(part + (quantity % 2))
        }
    }

    private mapGeolocationResponsesToGeolocationReferencesInfo(gelocationResponses: GeolocationResponse[]): GeolocationReferencesInfo[] {
        let responsesOk: GeolocationResponse[] = gelocationResponses.filter(resp => resp.status == "OK");
        let results: GeolocationResult[] = [];
        responsesOk.map(r => results = results.concat(r.results));
        return results.map(result => {
            let ref = new GeolocationReferencesInfo();
            ref.placeId = result.place_id;
            ref.address = result.formatted_address;
            ref.location = new Location();
            ref.location.lat = result.geometry.location.lat;
            ref.location.lng = result.geometry.location.lng
            ref.shorterDistances = [];
            ref.greatherDistances = [];
            ref.addresses = [];
            return ref;
        });
    }

    private mapGeolocationReferencesInfoToGeolocationInfo(geolocationRefencesInfo: GeolocationReferencesInfo): GeolocationInfo {
        const geolocationInfo = new GeolocationInfo();
        geolocationInfo.placeId = geolocationRefencesInfo.placeId;
        geolocationInfo.address = geolocationRefencesInfo.address;
        geolocationInfo.location = geolocationRefencesInfo.location;
        return geolocationInfo;
    }

    private calculateEuclidianDistance(refA: GeolocationReferencesInfo, refB: GeolocationReferencesInfo): number {
        return Math.sqrt(
            Math.pow(refA.location.lat - refB.location.lat, 2) +
            Math.pow(refA.location.lng - refB.location.lng, 2)
        );
    }

    private orderByDistanceAndTakeShorthers(reference: GeolocationReferencesInfo, shorterQuantity: number) {
        reference.shorterDistances = reference.addresses.sort((prev, actual) => {
            return prev.distance > actual.distance
                ? 1 : (prev.distance < actual.distance)
                    ? -1 : 0;
        }).slice(0, shorterQuantity);
    }

    private orderByDescendingDistanceAndTakeGreathers(reference: GeolocationReferencesInfo, greatherQuantity: number) {
        reference.greatherDistances = reference.addresses.sort((prev, actual) => {
            return prev.distance > actual.distance
                ? -1 : (prev.distance < actual.distance)
                    ? 1 : 0;
        }).slice(0, greatherQuantity);
    }    
}
